refactor: swap client management

erkarl opened the pull request:
In this PR we move swap client related logic out of various places in the codebase to `SwapClientsManager`. Still WIP, but good for a concept review.

Closes #898 
Closes #879

Commit:
refactor: swap client management

sangaman commented on lib/swaps/SwapClientsManager.ts:
```diff
@@ -0,0 +1,150 @@
+import { SwapClient } from '../constants/enums';
+import OrderBookRepository from '../orderbook/OrderBookRepository';
+
+class SwapClientsManager {
```
Minor nit but I think "SwapClientManager" without the s sounds better.

sangaman commented on lib/service/Service.ts:
```diff
@@ -121,17 +119,14 @@ class Service extends EventEmitter {
-      if (!client) {
-        // TODO: throw an error here indicating that lnd is disabled for this currency
+      try {
+        const client = this.swapClientsManager.get(currency.toUpperCase());
```
I think best to call this `swapClient` just to be consistent with how its called elsewhere

sangaman commented on lib/service/Service.ts:
```diff
@@ -121,17 +119,14 @@ class Service extends EventEmitter {
+        const client = this.swapClientsManager.get(currency.toUpperCase());
+        const channelBalance = await client.channelBalance();
+        return channelBalance;
+      } catch (e) {
```
I think this may be better without the try/catch, because if we get an error on the call it will bubble up to the grpc layer and return a grpc error. With the try/catch, if there's an error we return balances of 0 which is not accurate. I'd leave that TODO in there (and this will remind me to fix it) or if you want you can have it throw an error rather than return 0 balances when there's no swap client.

sangaman replied:
Nice solution!

sangaman commented on lib/swaps/SwapClientsManager.ts:
```diff
@@ -0,0 +1,150 @@
+  public get(currency: string): BaseClient {
+    const swapClient = this.swapClients.get(currency);
+    if (!swapClient) {
+      throw new Error(`swapClient for currency ${currency} not found`);
```
Maybe this is where we can throw that error I was talking about in my last comment, and ideally make it a defined error type rather than a generic `Error` so we can easily check for it.

sangaman commented on lib/swaps/SwapClientsManager.ts:
```diff
@@ -0,0 +1,150 @@
+import OrderBookRepository from '../orderbook/OrderBookRepository';
+
+class SwapClientsManager {
+  private lndClients: { [currency: string]: LndClient | undefined } = {};
```
I think this would be cleaner as a `Map` type.

sangaman reviewed (commented):
Made a quick pass with a few comments. One feature that would be really nice to have, and maybe makes sense to implement here, is making the `LndClient` output which currency it's for every time it writes a log statement. That way we can easily tell which `LndClient` is printing to logs, often times right now there's no real way to tell.

Conceptually though I like this and the manager class approach is good i think.

erkarl commented:
Feedback applied. Ready for round 2. Before merging I definitely want to also add TypeDoc comments to `SwapClientManager`.

I included https://github.com/ExchangeUnion/xud/pull/941 in this PR because I was getting inconsistent results when running the simulation tests.

sangaman commented on lib/swaps/SwapClientManager.ts:
```diff
@@ -0,0 +1,224 @@
+   * @param currency a currency that the swap client is linked to.
+   * @returns swap client instance upon success, throws otherwise.
+   */
+  public get(currency: string): BaseClient {
```
Just for consistency's sake, we're using arrow functions (almost) everywhere else so I think it's better to use that here too unless you have a reason not to, `public get = (currency: string): BaseClient {`. Same for other functions in this class.

sangaman commented on lib/swaps/SwapClientManager.ts:
```diff
@@ -0,0 +1,224 @@
+class SwapClientManager {
+  /** A map between currencies and all swap clients */
+  private swapClients = new Map<string, BaseClient>();
+  private raidenClient!: RaidenClient;
```
I would either make this nullable with the `?` operator, or create the `RaidenClient` object in the constructor (so that indeed this value will never be undefined) and then initialize it inside the `init` function here and let it become disabled.

sangaman commented on lib/swaps/SwapClientManager.ts:
```diff
@@ -0,0 +1,224 @@
+   * @param currency a currency that should be unlinked from a swap client.
+   * @returns Nothing upon success, throws otherwise.
+   */
+  public remove(currencyId: string): void {
```
To match the @param comment above (and how we're usually calling this throughout the codebase) I'd rename the param to `currency`.

sangaman commented on lib/swaps/SwapClientManager.ts:
```diff
@@ -0,0 +1,224 @@
+   */
+  public getLndPubKeys(): LndPubKeys {
+    const lndPubKeys: LndPubKeys = {};
+    for (const [currency, swapClient] of this.swapClients.entries()) {
```
What do you think about maintaining this list of pubkeys (similar to how we're maintaining the list of raiden currencies) rather than determining it dynamically when needed? I don't think it will be changing very often, so it might be ever so slightly clearer/more performant.

erkarl replied:
I'm not completely sold on the idea because of the added overhead of making sure that list stays up to date with swap clients' values. I added a TODO in the code for further consideration.

sangaman commented on lib/swaps/SwapClientManager.ts:
```diff
@@ -0,0 +1,224 @@
+  public getLndPubKeys(): LndPubKeys {
+    const lndPubKeys: LndPubKeys = {};
+    for (const [currency, swapClient] of this.swapClients.entries()) {
+      if (swapClient.type === SwapClient.Lnd) {
```
In several places we're checking `swapClient.type` and then casting. It'd be nice to create our own type checking functions that will automatically tell the compiler the type, like we for `isPeerOrder` and `isOwnOrder`

```typescript
export function isLndClient(swapClient: BaseClient): swapClient is LndClient {
  return (swapClient.type === SwapClient.Lnd);
}

export function isRaidenClient(swapClient: BaseClient): swapClient is RaidenClient {
  return (swapClient.type === SwapClient.Raiden);
}
```

erkarl replied:
Great idea!

sangaman commented on jest.config.js:
```diff
@@ -0,0 +1,7 @@
+module.exports = {
```
Mainly a question, but do you see value in keeping mocha and jest side by side? Or should we consider migrating to jest? Or maybe just the parts that are using stubbing, replacing sinon so to speak?

erkarl replied:
I'd do a gradual upgrade as the benefits of doing a full rewrite (less dev-dependencies) don't justify the cost (time consuming).

sangaman commented on lib/swaps/Swaps.ts:
```diff
@@ -256,7 +257,7 @@ class Swaps extends EventEmitter {
     const { rPreimage, rHash } = await generatePreimageAndHash();
     const peerPubKey = peer.nodePubKey!;
-    const swapClient = this.swapClients.get(currency);
+    const swapClient = this.swapClientManager.get(currency);
```
We probably want a try/catch here (rather than the if statement below) in case we have no swap client for this currency.

sangaman commented on lib/swaps/SwapClientManager.ts:
```diff
@@ -0,0 +1,224 @@
+    }
+    // bind event listeners
+    this.bind();
+    await Promise.all(initPromises);
```
It might make sense to remove swap clients that turned out to be disabled after the `initPromises` resolve, since we currently treat disabled clients as if they don't exist I believe.

erkarl replied:
Added a TODO for this since it's not a trivial change. We'd also need to readd the swap clients once their state becomes enabled.

sangaman commented on lib/swaps/Swaps.ts:
```diff
@@ -369,7 +370,7 @@ class Swaps extends EventEmitter {
     const { makerCurrency, makerAmount, takerCurrency, takerAmount } = Swaps.calculateMakerTakerAmounts(quantity, price, isBuy, requestBody.pairId);
 
-    const swapClient = this.swapClients.get(takerCurrency);
+    const swapClient = this.swapClientManager.get(takerCurrency);
```
Same here for try/catch if the swap client doesn't exist, replacing the if statement below.

sangaman commented on lib/swaps/Swaps.ts:
```diff
@@ -503,7 +504,7 @@ class Swaps extends EventEmitter {
 
-    const swapClient = this.swapClients.get(deal.makerCurrency);
+    const swapClient = this.swapClientManager.get(deal.makerCurrency);
     if (!swapClient) {
```
I think we can just remove this if statement (it will never be falsey), if don't have the swap client after we'd already checked it earlier then it's probably fine to let the exception bubble up.

sangaman commented on lib/swaps/Swaps.ts:
```diff
@@ -613,7 +614,7 @@ class Swaps extends EventEmitter {
       } else {
         // we need to get the preimage by making a payment
-        const swapClient = this.swapClients.get(currency);
+        const swapClient = this.swapClientManager.get(currency);
```
Same here for try/catch if the swap client doesn't exist, replacing the if statement below.

erkarl replied:
Why do we need to try catch here if the `SwapClientManager.get` will throw a valid error if the swap client doesn't exist?

sangaman commented on test/integration/Service.spec.ts:
```diff
@@ -35,10 +35,10 @@ describe('API Service', () => {
         },
         BTC: {
-          disable: true,
+          disable: false,
```
Why are these values changed?

erkarl replied:
With the new logic querying for a `swapClient` will throw if it's not found. The service spec relies on the the swap client to be available. With the simulation tests in place I'm beginning to question the need to have these kinds of acceptance tests 🤔 .

sangaman reviewed (commented):
Looking good, this is turning out to be a pretty big change but I think it'll be helpful to organize the logic handling swap clients before it grows even more complex.

sangaman commented:
Looks like the tests are failing due to lint errors, should be easy enough to fix.

erkarl commented:
@sangaman interesting - the same warnings that the linter was complaining about made my `npm run compile` fail on my other machine. Works fine on this machine, though.

sangaman commented:
Fixing the warnings made the compile fail? Or just that the cause of the warnings were compilation errors on one machine but not on the other?

It might make sense to update the linters, it's been a while.

erkarl commented:
> Fixing the warnings made the compile fail? Or just that the cause of the warnings were compilation errors on one machine but not on the other?

Those warnings presented as compilation errors. I cannot reproduce the situation after a clean `rm -Rf node_modules && npm i && npm run compile && npm run test` - everything passes. Very weird.

erkarl commented:
@sangaman I believe I've now addressed all your comments. I also cherry-picked https://github.com/ExchangeUnion/xud/pull/941 on top of this PR in order to make the simulation tests pass.

After applying changes to `Swaps.ts` I feel like making the `SwapClientManager.get()` to throw upon non-existent swap client might have been a mistake. I'm good with changing it back to `SwapClient || undefined` and handling the non-existent swap clients with the previous logic.

sangaman reviewed (commented):
Apologies, I wrote these comments a few days ago but forgot to actually hit submit so they were still pending.

erkarl commented on lib/Xud.ts:
```diff
@@ -112,32 +89,28 @@ class Xud extends EventEmitter {
+      if (!this.config.raiden.disable) {
+        // TODO: Should starting of the httpServer be handled as a callback from
+        // Raiden init/start/ready via the manager?
+        this.swapClientManager.get('WETH');
```
Good catch - the `get` call was not needed. I forgot it in from previous experiment.

erkarl commented on lib/Xud.ts:
```diff
@@ -112,32 +89,28 @@ class Xud extends EventEmitter {
       });
 
-      if (!this.raidenClient.isDisabled()) {
+      if (!this.config.raiden.disable) {
```
We no longer have `this.raidenClient` within the context of `Xud.ts`. I think it makes sense to enable the `httpServer` based on the configuration value of `raiden.disable` because if we're relying on the dynamic `raidenClient.isDisabled()` it could be in a state of trying to reconnect. However, once the reconnection happens we don't have logic to start the `httpServer`.

sangaman replied:
If it's trying to reconnect, it should have the `Disconnected` status instead of `Disabled`. `Disabled` means it's off and staying off, and although currently the only way raiden is disabled is via config, that could change one day (the lnd client can be disabled if we can't load the tls/macaroon files).

Could we expose the raiden status via the swap client manager? We might need it for other purposes down the road as well.

erkarl replied:
As discussed in private I ended up exposing the whole `raidenClient` instance and removing `SwapClientManager.getRaidenInfo/getRaidenAddress` as they were redundant.

erkarl commented on lib/swaps/Swaps.ts:
```diff
@@ -134,8 +134,12 @@ class Swaps extends EventEmitter {
     this.pool.on('packet.swapFailed', this.handleSwapFailed);
-    this.swapClients.forEach((swapClient, currency) => {
-      swapClient.on('htlcAccepted', async (rHash, amount) => {
+    // TODO: Made swapClientManager.swapClients public to temporarily/quickly
```
I removed the comment from the code and created https://github.com/ExchangeUnion/xud/issues/972 as I don't want to bloat this PR more than necessary.
